gradle/validation/validate-source-patterns.gradle

```diff
@@ -29,50 +33,127 @@ buildscript {
     group = 'Verification'
     description = 'Validate Source Patterns'
 
     // This task has no proper outputs.
     setupDummyOutputs(task)
 
-    sourceFiles = project.fileTree(project.rootDir) {
-      [
-        'java', 'jflex', 'py', 'pl', 'g4', 'jj', 'html', 'js',
-        'css', 'xml', 'xsl', 'vm', 'sh', 'cmd', 'bat', 'policy',
-        'properties', 'mdtext', 'groovy', 'gradle',
-        'template', 'adoc', 'json',
-      ].each{
-        include "lucene/**/*.${it}"
-        include "solr/**/*.${it}"
-        include "dev-tools/**/*.${it}"
-        include "gradle/**/*.${it}"
-        include "*.${it}"
+    sourceFiles = fileTree(projectDir) {
+      extensions.each{
+        include "**/*.${it}"
       }
-      // TODO: For now we don't scan txt / md files, so we
-      // check licenses in top-level folders separately:
-      include '*.txt'
-      include '*/*.txt'
-      include '*.md'
-      include '*/*.md'
-      // excludes:
+      
+      // Don't go into child projects (scanned separately).
+      childProjects.keySet().each{
+        exclude "${it}/**"
+      }
+
+      // default excludes.
+      exclude 'build/**'
+    }
+  }
+
+  // Add source validation to per-project checks as well.
+  check.dependsOn validateSourcePatterns
+}
+
+configure(project(':lucene:benchmark')) {
+  project.tasks.withType(ValidateSourcePatternsTask) {
+    sourceFiles.exclude 'temp/**'
+    sourceFiles.exclude 'work/**'
+  }
+}
+
+configure(project(':solr:core')) {
+  project.tasks.withType(ValidateSourcePatternsTask) {
+    sourceFiles.exclude 'src/**/CheckLoggingConfiguration.java'
+    sourceFiles.exclude 'src/test/org/apache/hadoop/**'
+  }
+}
+
+configure(rootProject) {
+  task validateRootAndOtherFiles(type: ValidateSourcePatternsTask) { task ->
+    // This task has no proper outputs.
+    setupDummyOutputs(task)
+
+    sourceFiles = fileTree(projectDir) {
+      extensions.each{
+        include "**/*.${it}"
+      }
+
+      // We do not scan for *.txt by default (broken files in subprojects),
+      // but in root we can do this).
+      include '**/*.txt'
+
+      // Don't go into child projects (scanned separately).
+      childProjects.keySet().each{
+        exclude "${it}/**"
+      }
+
+      // default excludes.
       exclude '**/build/**'
-      exclude '**/dist/**'
       exclude 'dev-tools/missing-doclet/src/**/*.java' // <-- TODO: remove once we allow "var" on master
-      exclude 'lucene/benchmark/work/**'
-      exclude 'lucene/benchmark/temp/**'
-      exclude '**/CheckLoggingConfiguration.java'
-      exclude 'solr/core/src/test/org/apache/hadoop/**'
-      exclude '**/validate-source-patterns.gradle' // ourselves :-)
+
+      // ourselves :-)
+      exclude 'gradle/validation/validate-source-patterns.gradle'
     }
   }
+
+  task validateSourcePatterns() {
+    group = 'Verification'
+    description = 'Validate Source Patterns'
+
+    // Make it depend on all so-named tasks from subprojects.
```

Comment:
(here).

Reply:
I also want this to be able to run it from root separately. If that's not needed and works automatically, remove it here.

I just made it consistent with other validation tasks

Reply:
Yes, it does work. Like I said - a task without a project qualifier is in fact an invocation of any such task in any subproject. So:
```
> gradlew -p lucene/analysis validateSourcePatterns  --console=plain
...
> Task :lucene:analysis:validateSourcePatterns UP-TO-DATE
> Task :lucene:analysis:phonetic:validateSourcePatterns UP-TO-DATE
> Task :lucene:analysis:opennlp:validateSourcePatterns UP-TO-DATE
> Task :lucene:analysis:stempel:validateSourcePatterns UP-TO-DATE
> Task :lucene:analysis:morfologik:validateSourcePatterns UP-TO-DATE
> Task :lucene:analysis:smartcn:validateSourcePatterns UP-TO-DATE
> Task :lucene:analysis:nori:validateSourcePatterns UP-TO-DATE
> Task :lucene:analysis:kuromoji:validateSourcePatterns UP-TO-DATE
> Task :lucene:analysis:icu:validateSourcePatterns UP-TO-DATE
> Task :lucene:analysis:common:validateSourcePatterns UP-TO-DATE
```